core/src/context.rs

```diff
@@ -128,83 +123,68 @@ impl TransformVariant {
+            package_manager: Arc::default(),
             resolver,
             #[cfg(feature = "native")]
             engine: EngineBuilder::new(Cranelift::new()).engine(),
             state: CompilationState::default(),
             filesystem: CoreFs::new(Arc::clone(&policy)),
             policy,
         };
-        ctx.load_default_packages()?;
+        #[cfg(feature = "native")]
+        ctx.package_manager
+            .lock()
+            .unwrap()
+            .load_default_packages(&ctx.engine)?;
+        #[cfg(not(feature = "native"))]
+        ctx.package_manager
+            .lock()
+            .unwrap()
+            .load_default_packages()?;
         Ok(ctx)
     }
 }
 
 impl<T, U> Context<T, U>
 where
     T: Resolve,
-    <T as Resolve>::Error: Error + 'static,
 {
-    pub(crate) fn import_missing_packages(&mut self, config: &Config) -> Result<(), CoreError> {
-        let missing: Vec<&str> = config
-            .imports
-            .iter()
-            .map(|i| i.name.as_str())
-            .chain(config.hides.iter().map(|h| h.name.as_str()))
-            .filter(|&name| {
-                !self.standard_packages.contains_key(name)
-                    && !self.external_packages.contains_key(name)
-            })
-            .collect();
-
-        // The .enumerate()-.map() may seem ugly but it is needed to be able to retrieve the
-        // package which errored (first). Essentially, instead of having just an dyn Error, we have
-        // (idx, dyn Error) where idx is the idx to the package (in missing) that errored
-        let resolves: Vec<Vec<u8>> = self
-            .resolver
-            .resolve_all(&missing)
-            .into_iter()
-            .enumerate()
-            .map(|(idx, pkg)| pkg.map_err(|e| (idx, e)))
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(|(idx, e)| CoreError::Resolve(missing[idx].to_string(), Box::new(e)))?;
-
-        let res: Result<Vec<()>, CoreError> = missing
-            .into_iter()
-            .zip(resolves.into_iter())
-            .map(|(name, data)| {
-                self.load_external_package(name, data.as_slice())
-                    .map(|_| ())
-            })
-            .collect();
-        res.map(|_| ())
-    }
-
     // This function configures the context with the given config, so that it is appropriate to
     // evaluate a document having that configuration with it. It also resolves packages if needed
-    pub(crate) fn configure(&mut self, config: Option<Config>) -> Result<(), CoreError> {
+    // If this returns "true", it had everything it needed to compile, if "false" it is waiting for
+    // more packages
+    pub(crate) fn configure(&mut self, config: Option<Config>) -> Result<bool, Vec<CoreError>> {
         let config = config.unwrap_or_default();
-        self.import_missing_packages(&config)?;
-        self.expose_transforms(config.try_into()?)?;
-        Ok(())
+        let mut lock = self.package_manager.lock().unwrap();
```

Comment:
package_store is a bad name since it isn't a package_store, it is a guard holding the lock to the store. Since we do `drop(lock)`, it is clear that we aren't dropping the store but just the lock on it; doing `drop(package_store)` isn't good since it looks like we are dropping the store itself which we aren't doing. Changed the name to `store_guard`.